zhusuan/distributions/special.py

```diff
@@ -0,0 +1,144 @@
+    """
+
+    def __init__(self, shape, dtype,
+                 group_ndims=0,
+                 is_continuous=None,
+                 **kwargs):
+        try:
+            self.static_shape = tf.TensorShape(shape[0])
+        except Exception:
+            self.static_shape = tf.TensorShape(tf.Dimension(None))
+        for s in shape[1:]:
+            try:
+                self.static_shape = self.static_shape.concatenate(tf.TensorShape(s))
+            except Exception:
+                self.static_shape = self.static_shape.concatenate(tf.Dimension(None))
+
+        if dtype is None:
+            dtype = tf.float32
+        assert_same_float_and_int_dtype([], dtype)
+
+        if is_continuous is None:
+            is_continuous = dtype.is_floating
+
+        super(Empirical, self).__init__(
+            dtype=dtype,
+            param_dtype=None,
+            is_continuous=is_continuous,
+            is_reparameterized=False,
+            group_ndims=group_ndims,
+            **kwargs)
+
+    def _value_shape(self):
+        return tf.constant([], dtype=tf.int32)
+
+    def _get_value_shape(self):
+        return tf.TensorShape([])
+
+    def _batch_shape(self):
+        raise ValueError("The Empirical distribution has no dynamic batch shape.")
+
+    def _get_batch_shape(self):
+        return self.static_shape
+
+    def _sample(self, n_samples):
+        raise ValueError("You can not sample from an Empirical distribution.")
+
+    def _log_prob(self, given):
+        raise ValueError("An empirical distribution has no log-probability measure.")
+
+    def _prob(self, given):
+        raise ValueError("An empirical distribution has no probability measure.")
+
+
+class Implicit(Distribution):
+    """
+    The class of Implicit distribution.
+    See :class:`~zhusuan.distributions.base.Implicit` for details.
+    :param name: A string. The name of the `StochasticTensor`. Must be unique
+        in the `BayesianNet` context.
+    :param implicit: A N-D (N >= 1) `float` Tensor
+    :param group_ndims: A 0-D `int32` Tensor representing the number of
+        dimensions in `batch_shape` (counted from the end) that are grouped
+        into a single event, so that their probabilities are calculated
+        together. Default is 0, which means a single value is an event.
+        See :class:`~zhusuan.distributions.base.Distribution` for more detailed
+        explanation.
+    """
+
+    def __init__(self,
+                 implicit,
+                 group_ndims=0,
+                 **kwargs):
+        self.implicit = implicit
+        super(Implicit, self).__init__(
+            dtype=implicit.dtype,
+            param_dtype=implicit.dtype,
+            is_continuous=implicit.dtype.is_floating,
+            group_ndims=group_ndims,
+            is_reparameterized=False,
+            **kwargs)
+
+    def _value_shape(self):
+        return tf.constant([], dtype=tf.int32)
```

Comment:
May not be univariate also. But I'm not sure how to deal with this though. Maybe an input argument `value_ndims`?